src/charm.py

```diff
@@ -230,32 +246,51 @@ def self_scraping_job(self):
 
         return [job]
 
+    def _update_blackbox_config_yaml_given_dict_from_relation(self, modules: dict) -> None:
+        """Update the blackbox config yaml given a dict of modules defined in relation.
+
+        This function takes the dict of modules from the BlackboxExporterRequirer and
+        updates the config file with the required modules, only if they do not exist already
+        in the config file.
+
+        Raises:
+            yaml.YAMLError: If there is an error in the YAML formatting or parsing.
+            TypeError: If type is not a string, as `yaml.safe_load` requires a string input.
+        """
+        if not modules:
+            return
+
+        config_file_data = self.container.pull(self._config_path).read()
+
+        if not config_file_data:
+            return
+
+        config_data = yaml.safe_load(config_file_data)
+
+        if "modules" not in config_data:
+            config_data["modules"] = {}
+
+        for module_name, module_data in modules.items():
+            # Add modules from relation data only if they do not exist in the config file
```

Comment:
The membership check `module_name not in config_data["modules"]` is not a diff; what if it is there but different?
Wdyt about rendering the config dict, and then comparing to what's in the file, and if it's different then write.

Reply:
that makes sense, changed in 8eaaac8170f5fdc10c1d2483b154b262fa145ead

Reply:
This works now, but if the charm that defines the module deletes one, there is currently no way of deleting modules. One way could be keeping a copy of the original config file somewhere, and rendering module holistically every time.

Reply:
See [comment](https://github.com/canonical/blackbox-exporter-k8s-operator/pull/43#issuecomment-2618371884)